pkg/util/pki/csr_test.go

```diff
@@ -553,3 +556,126 @@ func Test_buildKeyUsagesExtensionsForCertificate(t *testing.T) {
+
+	rootTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(0),
+		Subject: pkix.Name{
+			CommonName: "testing-root",
+		},
+		PublicKey: rootPrivKey.Public(),
+		IsCA:      true,
+	}
+
+	rootPem, rootCert, err := SignCertificate(rootTemplate, rootTemplate, rootTemplate.PublicKey, rootPrivKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	interPrivKey, err := GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	interTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: "testing-intermediate",
+		},
+		PublicKey: interPrivKey.Public(),
+		IsCA:      true,
+	}
+
+	interPem, interCert, err := SignCertificate(interTemplate, rootCert, interTemplate.PublicKey, rootPrivKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	leafPrivKey, err := GenerateRSAPrivateKey(2048)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	leafTemplate := &x509.Certificate{
+		SerialNumber: big.NewInt(2),
+		Subject: pkix.Name{
+			CommonName: "testing-leaf",
+		},
+		PublicKey: leafPrivKey.Public(),
+	}
+
+	leafPem, _, err := SignCertificate(leafTemplate, interCert, leafTemplate.PublicKey, interPrivKey)
+	if err != nil {
+		t.Error(err)
+		t.FailNow()
+	}
+
+	tests := []struct {
+		name              string
+		caCerts           []*x509.Certificate
+		caKey             crypto.Signer
+		template          *x509.Certificate
+		expectedCertPem   []byte
+		expectedCaCertPem []byte
+		wantErr           bool
+	}{
+		{
+			name:              "Sign intermediate template",
+			caCerts:           []*x509.Certificate{rootCert},
+			caKey:             rootPrivKey,
+			template:          interTemplate,
+			expectedCertPem:   interPem,
+			expectedCaCertPem: rootPem,
+			wantErr:           false,
+		},
+		{
+			name:              "Sign leaf template",
+			caCerts:           []*x509.Certificate{interCert, rootCert},
+			caKey:             interPrivKey,
+			template:          leafTemplate,
+			expectedCertPem:   append(leafPem, interPem...),
+			expectedCaCertPem: rootPem,
+			wantErr:           false,
+		},
+		{
+			name:              "Sign leaf template no root",
+			caCerts:           []*x509.Certificate{interCert},
+			caKey:             interPrivKey,
+			template:          leafTemplate,
+			expectedCertPem:   append(leafPem, interPem...),
+			expectedCaCertPem: interPem,
+			wantErr:           false,
+		},
+		{
+			name:     "Error on no CA",
+			caKey:    rootPrivKey,
+			template: rootTemplate,
+			wantErr:  true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			actualCertPem, actualCaCertPem, err := SignCSRTemplate(tt.caCerts, tt.caKey, tt.template)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("TestSignCSRTemplate() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !bytes.Equal(actualCertPem, tt.expectedCertPem) {
+				t.Errorf("TestSignCSRTemplate() certPem got = %v, want %v", actualCertPem, tt.expectedCertPem)
+			}
```

Comment:
I remember having a lot of trouble with finding out why the two PEM bytes (got, want) were different. I can't find exactly where I did that previously, but it looks like this:

```suggestion
			if !assert.Equal(t, tt.expectedCertPem, actualCertPem) {
				expected, _ := pki.DecodeX509CertificateBytes(tt.expectedCertPem)
				actualStr, _ := pki.DecodeX509CertificateBytes(actualCertPem)
				assert.Equal(t, expected, actual)
			}
```